File: tobas_setup_assistant/src/tobas_setup_assistant/parameter_getters/combo_box.py
from overrides import override
from typing import List, Optional
from PyQt5.QtCore import pyqtSignal, pyqtSlot

from tobas_rqt_tools.widgets import ComboBox

from .base import ParamGetterWidget


class ParamGetterWidget_ComboBox(ParamGetterWidget[str]):
    index_changed = pyqtSignal(int)
    text_changed = pyqtSignal(str)

    def __init__(
        self,
        param_name: str,
        description_text: Optional[str] = None,
        choices: List[str] = [],
        default: Optional[str] = None,
    ) -> None:
        super().__init__(param_name, description_text)

        self._box = ComboBox()
        self._rows.addWidget(self._box)

        self._box.addItems(choices)

        if default is not None:
            self._box.setCurrentText(default)

        self._box.currentIndexChanged.connect(self._on_index_changed)
        self._box.currentTextChanged.connect(self._on_text_changed)

    @override
    def get(self) -> str:
        return self._box.currentText()

    @override
    def set(self, src: str) -> None:
        self._box.setCurrentText(src)

    def cur_index(self) -> int:
        return self._box.currentIndex()

    def add_choices(self, items: List[str]) -> None:
        self._box.addItems(items)

    def set_choices(self, items: List[str]) -> None:
        self._box.clear()
        self.add_choices(items)

    @pyqtSlot(int)
    def _on_index_changed(self, idx: int) -> None:
        self.index_changed.emit(idx)

    @pyqtSlot(str)
    def _on_text_changed(self, text: str) -> None:
        self.text_changed.emit(text)

File: tobas_gui/tobas_setup_assistant/src/param_getters/combo_box.cpp
#include "tobas_setup_assistant/param_getters/combo_box.hpp"

namespace gui
{
namespace setup_assistant
{
ParamGetterWidget_ComboBox::ParamGetterWidget_ComboBox(
  const QString& param_name,
  const QString& description_text,
  const QStringList& choices,
  const QString& _default)
  : super(param_name, description_text)
{
  box_ = new qt::ComboBox();
  rows_->addWidget(box_);

  box_->addItems(choices);

  if (!_default.isEmpty())
    box_->setCurrentText(_default);

  connect(box_, SIGNAL(qt::ComboBox::currentIndexChanged(int)), this, SLOT(onIndexChanged(int)));
  connect(box_, SIGNAL(qt::ComboBox::currentTextChanged(const QString&)), this, SLOT(onTextChanged(const QString&)));
}

QString ParamGetterWidget_ComboBox::get() const
{
  return box_->currentText();
}

bool ParamGetterWidget_ComboBox::set(const QString& src)
{
  if (!box_->contains(src))
    return false;

  box_->setCurrentText(src);
  return true;
}

int ParamGetterWidget_ComboBox::currentIndex() const
{
  return box_->currentIndex();
}

void ParamGetterWidget_ComboBox::addChoices(const QStringList& items)
{
  box_->addItems(items);
}

void ParamGetterWidget_ComboBox::setChoices(const QStringList& items)
{
  box_->clear();
  addChoices(items);
}

void ParamGetterWidget_ComboBox::onIndexChanged(int index)
{
  Q_EMIT indexChanged(index);
}

void ParamGetterWidget_ComboBox::onTextChanged(const QString& text)
{
  Q_EMIT textChanged(text);
}
}  // namespace setup_assistant
}  // namespace gui
